File: app/src/main/java/com/umldesigner/infrastructure/uml/utils/ApiUtils.java
package com.umldesigner.infrastructure.uml.utils;

import android.content.Context;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.toolbox.Volley;
import com.umldesigner.infrastructure.uml.error.api.ApiErrorListener;

import lombok.AccessLevel;
import lombok.Getter;

/**
 * singleton which holds api related things
 */
public class ApiUtils {
    @Getter(AccessLevel.NONE)
    private static ApiUtils instance;
    /**
     * ip of the api
     */
    public static final String IP = "http://192.168.1.9:8080";
    
    /**
     * request queue for the api calls
     */
    private final RequestQueue requestQueue;
    
    /**
     * error listener for the api calls
     */
    @Getter
    private final Response.ErrorListener errorListener;
    
    private ApiUtils(Context context) {
        requestQueue = Volley.newRequestQueue(context);
        errorListener = new ApiErrorListener(context);
    }
    
    public static ApiUtils getInstance(Context context) {
        if(instance == null){
            instance = new ApiUtils(context);
        }
        return instance;
    }
    
    /**
     * adds a request to the queue
     * @param request the request that needs to be added to the queue
     */
    public void addRequest(Request<?> request){
       requestQueue.add(request);
    }
}

File: app/src/main/java/com/umldesigner/infrastructure/uml/utils/ApiUtils.kt
package com.umldesigner.infrastructure.uml.utils

import android.content.Context
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.Volley
import com.umldesigner.infrastructure.uml.error.api.ApiErrorListener
import lombok.AccessLevel
import com.umldesigner.infrastructure.uml.utils.ApiUtils
import lombok.Getter

/**
 * singleton which holds api related things
 */
class ApiUtils private constructor(context: Context) {
    /**
     * request queue for the api calls
     */
    private val requestQueue: RequestQueue

    /**
     * error listener for the api calls
     */

    val errorListener: Response.ErrorListener

    init {
        requestQueue = Volley.newRequestQueue(context)
        errorListener = ApiErrorListener(context)
    }

    /**
     * adds a request to the queue
     * @param request the request that needs to be added to the queue
     */
    fun addRequest(request: Request<*>?) {
        requestQueue.add(request)
    }

    companion object {
        private var instance: ApiUtils? = null

        /**
         * ip of the api
         */
        const val IP = "http://192.168.1.150:8080"
        @JvmStatic
        fun getInstance(context: Context): ApiUtils? {
            if (instance == null) {
                instance = ApiUtils(context)
            }
            return instance
        }
    }
}